// Copyright 2015, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

syntax = "proto3";

package coord438;

import "google/protobuf/timestamp.proto";

// The messenger service definition.
service CoordinatorService{

    rpc Login (Request) returns (Reply) {}
    rpc populateRoutingTable (Request) returns (Reply) {}
    rpc getAllClients (Request) returns (Reply) {}
    rpc getSlaveInfo (Request) returns (Reply) {}
    rpc getFSServerInfo (Request) returns (FSReply) {}
    rpc getServerInfo (Request) returns (FSReply) {}
    // Bidirectional Communication to check on server interaction
    rpc ServerCommunicate (stream HeartBeat) returns (stream HeartBeat) {}
 
}

enum ServerType{
    MASTER = 0;
    SLAVE = 1;
    SYNCHRONIZER = 3;
    COORDINATOR = 4;
}

enum RequesterType{
    CLIENT = 0;
    SERVER = 1;
}

message Request{
    // type of requester
    RequesterType requester = 1;
    // message with their port number
    string port_number = 2;
    // the id of the request
    int32 id = 3;
    // the type of server
    string server_type = 4;

    string all_clients_request = 5;
}

message Reply{
    string msg = 1;
    string port_number = 2;
    string server_type = 3;
    string server_id = 4;
    string all_clients_reply = 5;
    string slave_port = 6;
}

message HeartBeat{
    // id of the server
    string server_id = 1;
    // type of the server
    string s_type = 2;
    //Time the message was sent
    google.protobuf.Timestamp timestamp = 3;
}

message SlaveRequest{
    // id of the sever
    string server_id = 1;

}

message SlaveReply{
    string all_clients_for_cluster = 1;
    string followers = 2;
}

message FSReply{
    string port_number_1 = 1;
    int32 id_1 = 2;
    string port_number_2 = 3;
    int32 id_2 = 4;
}